### ISSUE TITLE
Dependency Injection (Guice) not working in Extractors / Filters ?

### ISSUE THREAD

**Issue by danielsawan** [ninjaframework/ninja#360]
The issue description is here : https://groups.google.com/forum/#!msg/ninja-framework/7j6GvCxshJ8/4NOlMIq229sJ

And to answer Raphael question "Can you inject UserDao in eg controller 
classes?" Yes it perfectly work in controllers.

**Comment by danielsawan** [comment]
Maybe it it because of the implementation "implements ArgumentExtractor<User>" . I don't know how Guice work but the only différence between Controller is this implementation.

**Comment by raphaelbauer** [comment]
No - it's not a Guice problem - it's very likely a bug in Ninja. 

An obvious workaround would be a simple component that can transform your context into the user. Plain simple java class (with Injects). That way the argument extractor bug won't hurt.

**Comment by metacity** [comment]
I can confirm this, but it only seems to manifest when annotating controller method parameters. When injecting the extractor into a filter class, for example, the field-injected fields in the extractor class seem to work fine.

**Comment by danielsawan** [comment]
can you try into Extractors ?

**Comment by metacity** [comment]
@danielsawan Sorry, I don't quite understand what you mean by that. And what Raphael suggested is quite straight-forward to do with zero modifications to your extractor class: inject the extractor to a controller and manually call its `extract(Context)` method in the controller methods. Even though you'll lose some of that Ninja magic, the following should work:

``` java
@Singleton
public class AppController {

    @Inject
    LoggedInUserExtractor userExtractor;

    public Result index(Context context) {
        User loggedInUser = userExtractor.extract(context);
        // ...
    }
}
```

I think field injection in filters worked ok.

**Comment by danielsawan** [comment]
Yes the injection of DAO seems to work in Filters but not in extractor. The only way i found to make it work easily in extrators : 

```
//  @Inject
    UserDao userDao;

    @Inject
    public LoggedInUserExtractor(UserDao userDao) {
        this.userDao = userDao;
    }
```

It is not a big deal by the way but still, it is not working as intended ;)

**Comment by lishid** [comment]
Jumping onto the same track here, I found this bug as well.

After tracing through Ninja's code, I ended up in the file `ControllerMethodInvoker.java`.

The Extractor is constructed through `getArgumentExtractor`, which calls `instantiateComponent`. A quick look at the method reveals that it does NOT use the injector if
- your Extractor has an empty constructor
- your Extractor has a constructor that takes a single parameter: the annotation type
- your Extractor has a constructor that takes a single parameter: a java.lang.Class object

Otherwise, the injector is used and you get a proper injection environment.

I would propose a fix to simply use the injector for every case, but there I'm not familiar with all the code around this so it might break some other things that rely on these simple constructors. The workaround is to create a @Inject constructor in your Extract class that takes any injected class (such as Context) other than the two mentioned above.

**Comment by raphaelbauer** [comment]
I guess instantiateComponent should just use guice like that:

```
private static <T> T instantiateComponent(Class<? extends T> argumentExtractor,
        final Annotation annotation, final Class<?> paramType,
        Injector injector) {
    return injector.createChildInjector(new AbstractModule() {
        @Override
        protected void configure() {
            bind((Class<Annotation>) annotation.annotationType()).toInstance(annotation);
            bind(ArgumentClassHolder.class).toInstance(new ArgumentClassHolder(paramType));
        }
    }).getInstance(argumentExtractor);
}
```

would you be willing to contribute some test cases?

**Comment by lishid** [comment]
Yeah, I can give it a go.

EDIT: Might take some time before I can get to it but I will once I'm migrated over.

**Comment by jlannoy** [comment]
@raphaelbauer I'm coming back on this bugfix.

1) Your solution will work but will require that each existing Extractor with a constructor got it preceded by an inject annotation. Easy to fix for Ninja's one, but user-made ones will need a manual update. I can document it on the upgrade guide, but there is now other way to do it properly (no deprectation possible here).

2) It also means that each time you need a ParamExtractor, instead of instantiating it directly (currently, a `newInstance()` inside the `instantiateComponent` method), it will required a Guice child injector... I'm not sure of the performance impact it can have.

Your thoughts?
